File: js/test/probe.mjs
// Adapter only: all examples and expected results live in examples/official.json.
import {createInterface} from 'node:readline';
import {readFileSync} from 'node:fs';
import {parseBytes, stringify, Value, ParseError} from '../index.js';

function tree(v) {
  switch (v.kind) {
    case 'object': return ['object', Array.from(v.members, ([key, value]) => [key, tree(value)])];
    case 'array': return ['array', v.items.map(tree)];
    case 'string': return ['string', v.stringValue()];
    case 'number': return ['number', v.numberLiteral()];
    case 'boolean': return ['boolean', v.booleanValue()];
    default: return ['null'];
  }
}
function rebuild(v) {
  if (v.kind === 'object') return Value.object([
    ...v.keys.map(key => [key, Value.null()]),
    ...v.keys.toReversed().map(key => [key, rebuild(v.get(key.stringValue()))]),
  ]);
  if (v.kind === 'array') return Value.array(v.items.map(rebuild));
  return v;
}
for await (const path of createInterface({input: process.stdin, crlfDelay: Infinity})) {
  const bytes = readFileSync(path);
  let value;
  try { value = parseBytes(bytes); }
  catch (error) {
    if (!(error instanceof ParseError)) throw error;
    console.log('{"ok":false}'); continue;
  }
  console.log(JSON.stringify({ok:true, raw:value.raw, serialized:stringify(value), compact:stringify(value,{compact:true}),
    tree:tree(value), rebuilt:stringify(rebuild(value),{compact:true})}));
}

File: rust/examples/probe.rs
// Adapter only. The shared Python verifier owns examples and expectations.
use ordered_json::{parse_bytes, stringify, Kind, OrderedMap, Value};
use std::{
    fs,
    io::{self, BufRead},
};

fn quote(s: &str) -> String {
    Value::string(s).raw().to_owned()
}
fn text(v: &Value) -> String {
    Value::from_units(v.string_units().unwrap())
        .raw()
        .to_owned()
}
fn tree(v: &Value) -> String {
    match v.kind() {
        Kind::Object => format!(
            "[\"object\",[{}]]",
            v.members()
                .unwrap()
                .iter()
                .map(|(key, value)| format!("[{},{}]", text(key), tree(value)))
                .collect::<Vec<_>>()
                .join(",")
        ),
        Kind::Array => format!(
            "[\"array\",[{}]]",
            v.items()
                .unwrap()
                .iter()
                .map(tree)
                .collect::<Vec<_>>()
                .join(",")
        ),
        Kind::String => format!("[\"string\",{}]", text(v)),
        Kind::Number => format!("[\"number\",{}]", quote(v.number_literal().unwrap())),
        Kind::Boolean => format!("[\"boolean\",{}]", v.boolean_value().unwrap()),
        Kind::Null => "[\"null\"]".into(),
    }
}
fn rebuild(v: &Value) -> Value {
    match v.kind() {
        Kind::Object => {
            let mut members = OrderedMap::new();
            for (key, _) in v.members().unwrap().iter() {
                members.insert(key.clone(), Value::null()).unwrap();
            }
            for (key, _) in v.members().unwrap().iter().rev() {
                members
                    .insert(
                        key.clone(),
                        rebuild(v.get_units(key.string_units().unwrap()).unwrap()),
                    )
                    .unwrap();
            }
            Value::object(&members).unwrap()
        }
        Kind::Array => {
            Value::array(&v.items().unwrap().iter().map(rebuild).collect::<Vec<_>>()).unwrap()
        }
        _ => v.clone(),
    }
}
fn main() {
    for line in io::stdin().lock().lines() {
        let bytes = fs::read(line.unwrap()).expect("cannot read document");
        match parse_bytes(&bytes) {
            Err(error) => println!(
                "{{\"ok\":false,\"offset\":{},\"unit\":\"byte\",\"kind\":{}}}",
                error.offset,
                quote(error.kind())
            ),
            Ok(v) => {
                let serialized = stringify(&v);
                let roundtrip = parse_bytes(serialized.as_bytes()).unwrap();
                let factory = Value::string("quote \" slash \\ line\n 한 🌍")
                    .raw()
                    .to_owned();
                println!(
                "{{\"ok\":true,\"raw\":{},\"serialized\":{},\"compact\":{},\"tree\":{},\"roundtrip\":{},\"roundtrip_tree\":{},\"rebuilt\":{},\"factory\":{}}}",
                quote(v.raw()),
                quote(&serialized),
                quote(&v.compact()),
                tree(&v),
                quote(&stringify(&roundtrip)),
                tree(&roundtrip),
                quote(&rebuild(&v).compact()),
                quote(&factory)
                )
            }
        }
    }
}
